libs/parallelism/algorithms/include/hpx/parallel/algorithms/uninitialized_copy.hpp

```diff
@@ -129,118 +338,169 @@ namespace hpx { namespace parallel { inline namespace v1 {
+                    }
+                    return util::in_out_result<InIter1, FwdIter2>{
+                        first, current};
+                }
+                catch (...)
+                {
+                    for (/* */; dest != current; ++dest)
+                    {
+                        (*dest).~value_type();
+                    }
+                    throw;
+                }
+            }
+
+            template <typename ExPolicy, typename Iter, typename Sent1,
+                typename FwdIter2, typename Sent2>
+            static typename util::detail::algorithm_result<ExPolicy,
+                util::in_out_result<Iter, FwdIter2>>::type
+            parallel(ExPolicy&& policy, Iter first, Sent1 last, FwdIter2 dest,
+                Sent2 last_d)
+            {
+                std::size_t dist1 = detail::distance(first, last);
+                std::size_t dist2 = detail::distance(dest, last_d);
+                std::size_t dist = dist1 <= dist2 ? dist1 : dist2;
+
+                return parallel_sequential_uninitialized_copy_n(
+                    std::forward<ExPolicy>(policy), first, dist, dest);
+            }
+        };
+        /// \endcond
+    }    // namespace detail
+
     /////////////////////////////////////////////////////////////////////////////
     // uninitialized_copy_n
     namespace detail {
         /// \cond NOINTERNAL
-        template <typename FwdIter2>
+        template <typename IterPair>
         struct uninitialized_copy_n
-          : public detail::algorithm<uninitialized_copy_n<FwdIter2>, FwdIter2>
+          : public detail::algorithm<uninitialized_copy_n<IterPair>, IterPair>
         {
             uninitialized_copy_n()
               : uninitialized_copy_n::algorithm("uninitialized_copy_n")
             {
             }
 
-            template <typename ExPolicy, typename InIter>
-            static FwdIter2 sequential(
+            template <typename ExPolicy, typename InIter, typename FwdIter2>
+            static util::in_out_result<InIter, FwdIter2> sequential(
                 ExPolicy, InIter first, std::size_t count, FwdIter2 dest)
             {
-                return std::uninitialized_copy_n(first, count, dest);
+                using value_type =
+                    typename std::iterator_traits<FwdIter2>::value_type;
+
+                FwdIter2 current = dest;
+                try
+                {
+                    for (/* */; count > 0; ++first, (void) ++current, --count)
+                    {
+                        ::new (std::addressof(*current)) value_type(*first);
+                    }
+                    return util::in_out_result<InIter, FwdIter2>{
+                        first, current};
+                }
+                catch (...)
+                {
+                    for (/* */; dest != current; ++dest)
+                    {
+                        (*dest).~value_type();
+                    }
+                    throw;
+                }
```

Comment:
Could this code be extracted into a separate function to avoid code duplication?

Reply:
Can this use `std_uninitialized_copy_sent` now as well? Or is that code too different from the other function?

Reply:
They differ in the number of parameters and the condition checked in the for loop. I guess we could do this https://gist.github.com/Jedi18/4b016ea0d857fe9b7911c5a8a85bce72

Reply:
Sounds good to me!

Reply:
Does this still need addressing?

Reply:
@Jedi18 ping? What's the status of this?

Reply:
It's done, I've seperated it out as requested.